src/ProjectTemplates/BlazorTemplates.Tests/BlazorServerTemplateTest.cs

```diff
@@ -107,84 +136,94 @@ public async Task BlazorServerTemplateWorks_IndividualAuth(bool useLocalDB)
                     aspNetProcess.Process.HasExited,
                     ErrorMessages.GetFailedProcessMessageOrEmpty("Run published project", Project, aspNetProcess.Process));
 
                 await aspNetProcess.AssertStatusCode("/", HttpStatusCode.OK, "text/html");
-                if (BrowserFixture.IsHostAutomationSupported())
+                if (Fixture.BrowserManager.IsAvailable(browserKind))
                 {
-                    aspNetProcess.VisitInBrowser(Browser);
-                    TestBasicNavigation();
+                    var page = await aspNetProcess.VisitInBrowserAsync(browser);
+                    await TestBasicNavigation(page);
+                    await page.CloseAsync();
+                }
+                else
+                {
+                    Assert.False(
+                        TestHelpers.TryValidateBrowserRequired(
+                            browserKind,
+                            isRequired: !Fixture.BrowserManager.IsExplicitlyDisabled(browserKind),
+                            out var errorMessage),
+                        errorMessage);
                 }
             }
         }
 
-        private void TestBasicNavigation()
+        private async Task TestBasicNavigation(IPage page)
         {
-            var retries = 3;
-            var connected = false;
-            do
-            {
-                try
-                {
-                    Browser.Contains("Information: WebSocket connected to",
-                        () => string.Join(Environment.NewLine, Browser.GetBrowserLogs(LogLevel.Info).Select(b => b.Message)));
-                    connected = true;
-                }
-                catch (TimeoutException) when(retries-- > 0)
-                {
-                    Browser.Navigate().Refresh();
-                }
-            } while (!connected && retries > 0);
+            var socket = BrowserContextInfo.Pages[page].WebSockets.SingleOrDefault() ??
+                (await page.WaitForEventAsync(PageEvent.WebSocket)).WebSocket;
 
+            await socket.WaitForEventAsync(WebSocketEvent.FrameReceived);
 
-            Browser.Exists(By.TagName("ul"));
+            await page.WaitForSelectorAsync("ul");
             // <title> element gets project ID injected into it during template execution
-            Browser.Equal(Project.ProjectName.Trim(), () => Browser.Title.Trim());
+            Assert.Equal(Project.ProjectName.Trim(), (await page.GetTitleAsync()).Trim());
 
             // Initially displays the home page
-            Browser.Equal("Hello, world!", () => Browser.FindElement(By.TagName("h1")).Text);
+            await page.WaitForSelectorAsync("h1 >> text=Hello, world!");
 
             // Can navigate to the counter page
-            Browser.Click(By.PartialLinkText("Counter"));
-            Browser.Contains("counter", () => Browser.Url);
-            Browser.Equal("Counter", () => Browser.FindElement(By.TagName("h1")).Text);
+            await Task.WhenAll(
```

Comment:
Doesn't this change the expectation from before? For e.g. if the initial page rendered a Counter (because we have a bug in our template or framework), this would now succeed whereas it would not have previously?

Reply:
Also, it's not clear what the concurrent execution is meant to solve? In the ordinary case, these should generally evaluate serially just because that's when items would be ready no?

Reply:
No, the `await page.WaitForSelectorAsync("h1 >> text=Hello, world!");` above makes sure that's not the case.

In this case it can be done sequentially, but by doing it this way has the benefit that we make sure that we don't miss any change if it happens fast after the interaction.

Reply:
> any change if it happens fast after the interaction.

Can you elaborate on this? Do we have a concurrency issue with our tests this is trying to fix?

Reply:
Perhaps there are certain kinds of changes that can be missed if they happen too soon (for example, a state that exists only briefly), but it doesn't seem like it applies to the assertions we're doing here, since the assertions are checking that the document reaches a desired state before we mutate the document further.

In general people think about test scripts as a sequence of actions and observations performed serially. It will be easier to reason about if we stick closer to that pattern rather than having higher-level constructs like parallelism.